services/121-service/src/payments/services/payments-execution.service.ts

```diff
@@ -66,20 +72,74 @@ export class PaymentsExecutionService {
         programFspConfigurationNames,
       );
 
+      // TODO these 2 actions will later be slit up. For now they are together.
+      await this.paymentEventsService.createEvent({
+        paymentId,
+        userId,
+        type: PaymentEvent.approved,
+      });
       await this.paymentEventsService.createEvent({
         paymentId,
         userId,
         type: PaymentEvent.started,
       });
 
+      // update all "included" transactions to 'approved' ..
+      const programFspConfigurationIdMap = new Map<number, number>(
+        transactionsPendingApproval.map((item) => [
+          item.id,
+          item.programFspConfigurationId!,
+        ]),
+      );
+      const transactionsToStart = transactionsPendingApproval.filter(
+        (t) =>
+          t.registration.registrationStatus === RegistrationStatusEnum.included,
+      );
+      if (transactionsToStart.length > 0) {
+        await this.transactionsService.saveTransactionProgressBulk({
+          newTransactionStatus: TransactionStatusEnum.approved,
+          transactionIds: transactionsToStart.map((t) => t.id),
+          description: TransactionEventDescription.approved,
+          type: TransactionEventType.approval,
+          userId,
+          programFspConfigurationIdMap,
+        });
+      }
+
+      // .. and all "non-included" transactions to failed
+      const transactionsToFail = transactionsPendingApproval.filter(
+        (t) =>
+          t.registration.registrationStatus !== RegistrationStatusEnum.included,
+      );
+      if (transactionsToFail.length > 0) {
+        await this.transactionsService.saveTransactionProgressBulk({
+          newTransactionStatus: TransactionStatusEnum.error,
+          transactionIds: transactionsToFail.map((t) => t.id),
+          description: TransactionEventDescription.notApproved,
+          type: TransactionEventType.approval,
+          userId,
+          programFspConfigurationIdMap,
+          errorMessages: new Map<number, string>(
+            transactionsToFail.map((t) => [
+              t.id,
+              'Registration did not have status included at the time of starting the payment',
+            ]),
+          ),
+        });
+      }
+
+      // Note: this includes also the failing transactions to be consistent with other failed transactions for now.
       await this.paymentsExecutionHelperService.updatePaymentCountAndSetToCompleted(
         {
-          registrationIds: transactionsToStart.map((t) => t.registrationId),
+          registrationIds: transactionsPendingApproval.map(
+            (t) => t.registrationId,
+          ),
           programId,
           userId,
         },
       );
```

Comment:
This method became more complex. I think we should move more to helper functions. It may be good pair program opportunity actually

Reply:
done